Spatial Transformer Network Ops using cuDNN

joaovictortr opened the pull request:
Goal: implement [Spatial Transformer Networks'](https://arxiv.org/abs/1506.02025) Ops using cuDNN

Issue: #5622 

Tasks:

- [x] Add Op to create and manage a spatial transformer network descriptor (cudnnSpatialTransformerDescriptor_t)
- [x] Implement Grid Generator Op
- [x] Implement Grid Sampler Op
- [x] Add gradients to Grid Generator Op (backpropagation)
- [x] Add gradients to Grid Sampler Op (backpropagation)
- [x] Document functions and classes in Theano's documentation

joaovictortr commented:
Hello @notoraptor , I've added this PR to keep track of Spatial Transformer Network's Ops development. I'll update it as the implementation evolves.

joaovictortr commented:
Hello @notoraptor @nouiz 

I've implemented the basic functionality for the spatial transformer (grid of coordinates and the sampler), but I have some doubts.

I'm assuming the memory layout is NCHW (number of images, channels, height, width), is that assumption correct?

Also, I'm not clear on how to implement the gradients, should they be in a separate class (e.g. DNN convolution) or in the L_Op method?

notoraptor commented:
Hello, @joaovictortr :

1. Yes, assume tensors are NCHW. You will just have to document this assumption.
2. It may be better if you follow the same structure as with cuDNN convolutions, ie. a separate Op for each gradient (GpuDnnConvGradW, GpuDnnConvGradI). Ultimately, these Ops will be used in grad() (or L_op()) method.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,186 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        return (super(GpuDnnSpatialTfDesc, self).c_code_cache_version(), version())
+
+
+class GpuDnnGridGeneratorOp(DnnBase):
```
Rename it to `GpuDnnGridGenerator`.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,186 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        pass
+
+
+class GpuDnnGridSamplerOp(DnnBase):
```
Rename it to `GpuDnnGridSampler`.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,186 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def __init__(self, dimensions, precision="float32"):
+        COp.__init__(self, ["c_code/spatialtf_desc.c"], "APPLY_SPECIFIC(spatialtf_desc)")
+
+        self.dimensions = dimensions if isinstance(dimensions, tuple) else tuple(dimensions)
```
You can just write `self.dimensions = tuple(dimensions)`. Python's `tuple()` should be smart enough to just return the tuple if it receives a tuple, and generate a tuple if it receives a non-tuple.

joaovictortr replied:
You're correct, I've fixed that as well

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,186 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+        # cuDNN supports only 2D transformations, therefore output tensor must
+        # not exceed 4 dimensions (width, height, num_feature_maps, num_images)
+        assert len(self.dimensions) <= 4
```
It seems you expect at least 2 dimensions (dim0, dim1) ? If so, also check that len(self.dimensions) >= 2.

joaovictortr replied:
Yes, their layout must include at least the width and height. I've added an assertion to guarantee that it has at least these two dimensions.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,186 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+            context_name = infer_context_name(desc, theta)
+        else:
+            context_name = infer_context_name(desc, theta, cx)
+
```
Don't forget to apply `as_gpuarray_variable` to inputs here.

joaovictortr replied:
I've added the `as_gpuarray_variable` to the inputs

joaovictortr commented:
Hello @notoraptor 

> Yes, assume tensors are NCHW. You will just have to document this assumption.

Using a real image example, I've found that the usual layout is NHWC, I've updated the code and documented the assumption. I'll also add it to the docstring later on.

> are some of these Ops already fully implemented ?

In principle yes, the forward spatial transformer is fully implemented. The localization network can be built using already existent components of Theano - it can be a CNN or fully-connected neural net, it just needs to have a regression layer at the output. Therefore, I've implemented the grid generator and sampler, using the corresponding functions from cuDNN.

However, when I have tried with a real image example, applying an identity transformation, I've only got noise as output. I'm currently investigating the issue. I've added the example I'm testing, please let me know if you have any ideas on this issue.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,189 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        beta = ensure_dt(beta, _zero, 'beta', img.dtype)
+
+        return Apply(self, [img, grid, grid_dimensions, desc, alpha, beta],
+                     [output.type()])
```
I think we could just use the same output type as `grid` input (or as `img` input?), and then we don't need to pass an `output` variable to `make_node()`.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,189 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    """
+        GPU spatial transformer using cuDNN from NVIDIA.
+    """
+    precision = theano.config.floatX
```
I think this line should not be here, as precision should be given as function arg.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,189 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    desc = GpuDnnSpatialTfDesc(grid_dims, precision)()
+
+    # Create grid dimensions variable
+    grid_dims_arr = np.asarray(list(grid_dims), dtype=np.int32)
```
grid_dims may be a Theano symbolic variable, isn't it ? So, apply a `np.asarray()` may fail. Just apply `as_tensor_variable()` as you do in next line.

joaovictortr commented on theano/gpuarray/c_code/spatialtf_grid.c:
```diff
@@ -0,0 +1,86 @@
+        return -1;
+    }
+
+    if ( NULL == *grid )
```
I've added a memset to reset the memory in those cases, as in the CTC case

notoraptor commented:
Hello, @joaovictortr . I can run your test, but I still don't find why resulted images are so noisy.

But just one remark: all other theano Ops expect inputs to be NCHW (or at least, N first, C second, and generally HW follows, e.g. for convolutions: http://deeplearning.net/software/theano_versions/dev/tutorial/conv_arithmetic.html#theano-terminology ), so I strongly suggest that the spatial transformation Ops also work with NCHW tensors, to keep a consistency across all Theano ops. We should better just mention in doc that we expect NCHW inputs, so that the user will have to swap dimensions if necessary before using images.

For your test, the image returned by SciPy is indeed in HWC format: just adapt it before using. With NumPy, you can transpose (-> CWH) and then tensor.swapawes(1, 2) (-> CHW).

joaovictortr commented on theano/gpuarray/c_code/spatialtf_grid.c:
```diff
@@ -0,0 +1,86 @@
+        return -1;
+    }
+
+    if ( PyArray_DIM( grid_dimensions, 0 ) < 3 )
```
Grid dimensions should be at least 3D (number of images, height, width), since gray-scale images do not have the number of channels dimension.

joaovictortr commented on theano/gpuarray/c_code/spatialtf_grid.c:
```diff
@@ -0,0 +1,86 @@
+    const void * theta_data = PyGpuArray_DEV_DATA( theta );
+    void * grid_data = PyGpuArray_DEV_DATA( *grid );
+
+    err = cudnnSpatialTfGridGeneratorForward( _handle, desc, theta_data, grid_data );
```
Added that check as well, I check if the inptus and outputs are all contiguous on the GPU

notoraptor commented on theano/gpuarray/c_code/spatialtf_sampler.c:
```diff
@@ -0,0 +1,198 @@
+        return -1;
+    }
+
+    if ( NULL == *output )
```
Sorry, I forget, but here also, output may not be null.

joaovictortr replied:
Do you mean I have to check if output is NULL before using it here?

notoraptor replied:
I mean, like any other Theano output variable, you should check if either it is null or if it is allocated with bad shapes, and reallocate if so (as you do with `grids` in another C code).

notoraptor replied:
Don't forget to check `output` here when it is not null, as when you use `theano_size_check` in `spatialtf_grid.c`.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,59 @@ def funcs(out, params):
+        plt.show()
+
+    topo = spatialtf_fn.maker.fgraph.toposort()
+    assert len([n for n in topo if isinstance(n.op, dnn.GpuDnnGridGeneratorOp)]) == 1
```
GpuDnnGridGeneratorOp -> GpuDnnGridGenerator

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,189 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, grid_dimensions, theta, desc, precision, cx=None):
```
I've implemented these suggestions, and fixed the parameter ordering in the C op as well.

Yes, `precision` should be the same cuDnnDataType_t as specified in the spatial transformer descriptor.

joaovictortr commented:
Hello @notoraptor ,

> But just one remark: all other theano Ops expect inputs to be NCHW (or at least, N first, C second, and generally HW follows, e.g. for convolutions: http://deeplearning.net/software/theano_versions/dev/tutorial/conv_arithmetic.html#theano-terminology ), so I strongly suggest that the spatial transformation Ops also assume NCHW tensors, to keep a consistency across all Theano ops. We should better just mention in doc that we expect NCHW inputs, so that the user will have to swap dimensions if necessary before using images.

I have changed the tensor format in the spatial transformer back to NCHW, as suggested.

> For your test, the image returned by SciPy is indeed in HWC format: just adapt it before using. With NumPy, you can transpose (-> CWH) and then tensor.swapawes(1, 2) (-> CHW).

In the test, I am now transposing the image to be in the proper tensor format.

I have also found the problem, I was setting the input tensor dimensions incorrectly in the sampler. Instead of using the input width and height, I was using the grid's width and height in the input tensor. Changing it to the input width and height, fixed the problem.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        self.precision = precision
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
```
As this `dnn_context()` has currently the same definition as in parent class `DnnBase`, you can remove it here.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        self.precision = precision
+
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
```
As this `dnn_context()` has currently the same definition as in parent class `DnnBase`, you can remove it here.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        grid_dimensions = as_tensor_variable(grid_dimensions)
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
```
From your C code, you expect theta to be float, half or double, but cudnnDataType may include `int8` or other types (e.g. in cuDNN V6). So here, check explicitely that theta.dtype is in ('float16', 'float32', 'float64').

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
+        assert theta.ndim == 3
```
Also check than grid_dimensions.ndim == 1.

notoraptor commented on theano/gpuarray/c_code/spatialtf_grid.c:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    if ( NULL == *grid ||
+         ! theano_size_check( *grid, 4, &(grid_dims[0]), (*grid)->ga.typecode ) )
```
You could directly write `grid_dims` instead of `&(grid_dims[0])`.

notoraptor commented on theano/gpuarray/c_code/spatialtf_grid.c:
```diff
@@ -0,0 +1,107 @@
+    }
+    else
+    {
+        GpuArray_memset( &( (*grid)->ga ), 0 );
```
NB: For info, do you know if grid memory need to be set to 0 ? Can you try your code with pygpu_empty at line 65 and with this line commented ? If it works, we may save time by avoiding some calls to memset() !

joaovictortr replied:
I've checked with a shearing matrix, which leaves some pixels of the output image uncovered, and they are correctly initialized to zero by cuDNN. In this case, I've used pygpu_empty and removed the memset calls.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+        if img.type.ndim != 4:
+            raise TypeError('img must be a 4D tensor')
+        if output.type.ndim != 4:
```
Output will have the same ndim as img, so you can remove this checking.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        if output.type.ndim != 4:
+            raise TypeError('output must be a 4D tensor')
+
+        if img.type.ndim != output.type.ndim:
```
You can remove this code also.

notoraptor commented on theano/gpuarray/c_code/spatialtf_sampler.c:
```diff
@@ -0,0 +1,220 @@
+    float af = alpha, bf = beta;
+    spatialtf_context_t spatialtf_ctx;
+    cudnnDataType_t dt;
+    // Number of color channels (feature maps) is the innermost dimension
```
Is that always true ?

joaovictortr replied:
That comment should have been removed, sorry about that. In this case, we assume that the width is the innermost dimension, because of the NCHW data layout.

notoraptor commented on theano/gpuarray/c_code/spatialtf_sampler.c:
```diff
@@ -0,0 +1,220 @@
+    }
+    else
+    {
+        GpuArray_memset( &( (*output)->ga ), 0 );
```
Here also, check that it works with just pygpu_empty and without memset. If so, we could skip calls to memset. If not we will keep the code as it is now.

joaovictortr replied:
It also works with `pygpu_empty` and without memset

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def dnn_context(self, node):
+        return node.outputs[0].type.context_name
+
+    def make_node(self, img, grid, grid_dimensions, desc,
```
Should `grid_dimensions` be the same as `grid`'s dimensions ? If so, we should avoid passing `grid_dimensions` to this `make_node()`, and just get grid dimensions in C code with `PyGpuArray_DIM(grid, 0)`, etc.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        pass
+
+
+def dnn_spatialtf(img, theta, grid_dims, alpha=None, beta=None, precision=None):
```
`precision` seems to be in fact the dtype of output. And I was thinking that the dtype of output should just be the dtype of input. What do you think ?

In any case, rename it here to dtype, output_dtype (maybe the best name ?) or output_type.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    img = gpu_contiguous(img)
+    theta = gpu_contiguous(theta)
+
+    precision = get_precision(precision, [img, theta])
```
If output dtype is None, it should be the input (img) dtype by default.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,64 @@ def funcs(out, params):
+
+    img_out = np.asarray(result, dtype=np.float32)
+
+    print(img_out.shape)
```
To be removed (in final test code). We prefer avoid (too long) printing.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,64 @@ def funcs(out, params):
+    img_out = img_out.astype(dtype=np.uint8)
+    img_out = np.transpose(img_out, axes=(0, 2, 3, 1))
+
+    for i in range(len(img_out)):
```
To be removed.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,64 @@ def funcs(out, params):
+        print(img_out[i])
+
+    import matplotlib.pyplot as plt
+    for img_idx in range(len(img_out)):
```
To be removed also.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,64 @@ def funcs(out, params):
+
+    topo = spatialtf_fn.maker.fgraph.toposort()
+    assert len([n for n in topo if isinstance(n.op, dnn.GpuDnnGridGenerator)]) == 1
+    assert len([n for n in topo if isinstance(n.op, dnn.GpuDnnGridSampler)]) == 1
```
We should remove printing and images plotting from the final code of this test. Do you think here is another way to check if the output images are indeed correctly computed ?

joaovictortr replied:
Definitely, as it also introduces additional dependencies (e.g. `scipy`). Yes, there is a [CPU implementation](http://lasagne.readthedocs.io/en/latest/modules/layers/special.html?highlight=SpatialTransformer) using Theano in Lasagne. I will try to use it to check the computed values.

Do you think I should also add (with proper credits to the authors) the CPU spatial transformer from Lasagne into Theano?

notoraptor replied:
It might be a good idea ! But as it was not planned, we should better end with GPU implementation first, and perhaps with GSoC second. If there is time (or if you're motivated!), you could add CPU implementation later in another PR!

What do you think, @nouiz ?

notoraptor replied:
NB: About that, let's end with CTC pull request and this PR for GPU Spatial Transformation.

After that, we will discuss if your last part of GSoC will be to implement either CPU version for spatial transformation, or something else.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,191 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        COp.__init__(self, ["c_code/spatialtf_desc.c"], "APPLY_SPECIFIC(spatialtf_desc)")
+
+        # dimensions must have at least width and height
+        assert len(dimensions) >= 2
```
Absolutely, even a simple `reshape((1, 1, width, height))` is enough to provide the proper data layout. I've modified the assert to check if the img has four dimensions.

nouiz commented:
I killed travis to help fix travis timeout on the master branch faster

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def do_constant_folding(self, node):
+        return False
+
+    def __init__(self, dimensions, dtype="float32"):
```
I think we should better use `theano.config.floatX` as default value for dtype.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    """
+    This Op builds a spatial transformer descriptor for use in spatial transformer network
+    operations.
```
Add doc for op attributes (dimensions and dtype), and describe the expected format for dimensions (NCHW),

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+
+        assert theta.dtype in ('float16', 'float32', 'float64')
+        assert cudnn.cudnnDataType_t.has_alias(theta.dtype)
```
We can remove this line, as previous line already checks `theta.dtype` with a fixed set of supported types.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+def dnn_spatialtf(img, theta, grid_dims, alpha=None, beta=None, dtype=None):
+    """
+        GPU spatial transformer using cuDNN from NVIDIA.
```
Could you extend this doc, to explain a little what is computed here, and to add doc about function parameters ?

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2732,6 +2732,176 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    # img is a 4D tensor with shape: (num_images, num_channels, width, height)
+    assert img.ndim == 4
+    # Grid dimensions must be a 4-dimensional tuple
```
"a 4-length tuple" instead, I think. Also, I think you should allow iterable objects (tuples, or lists).

```python
# instead of `assert isinstance(grid_dims, tuple)`:
grid_dims = tuple(grid_dims)
```

notoraptor commented on theano/gpuarray/c_code/spatialtf_sampler.c:
```diff
@@ -0,0 +1,215 @@
+        cuda_exit( gpu_ctx->ctx );
+
+        PyErr_Format( PyExc_RuntimeError,
+                      "Could not create xdesc: %s",
```
`ydesc` here, not `xdesc`.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2131,3 +2131,89 @@ def funcs(out, params):
+    def normalize_input(input):
+        # Scale input from [0, 255] to [0, 2]
+        scale_factor = 2 ** -7  # 1/128
+        input *= scale_factor
```
Not important, but for better precision, I think you should directly apply division, instead of compute scale_factor and then apply a multiplication:
```python
input /= 128.0
```

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2800,6 +2800,188 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        pass
+
+
+def dnn_spatialtf(inp, theta, downsampling_factor=1, alpha=None, beta=None, dtype=theano.config.floatX):
```
I see you are changing the whole interface for spatial transformation by adding a `downsampling` factor. But from what I understand, an user may want to do a spatial transformation that is different for width and for height, right ? So, indeed, number of images and number of channels per image should remain the same, but there may be a new independent width and a new independent height, isn't it ?

If so, using one `downsampling` factor is a too limited strategy. I think you should better directly use params like `new_width` and `new_height`, to allow the user to explicitely specify the dimensions of the transformed images.

Or, maybe better, use params like `scale_width` and `scale_height` to allow user to explicitely specify a downsample/upsample factor for each dimension, so that:
```python
new_width = image_width * scale_width
new_height = image_height * scale_height
```
In such case (and also with current downsampling factor), I think a round strategy (probably round half up) should be applyed to get new width and new height as integers, isn't it ?

joaovictortr replied:
I agree, while having the user specify the number of images and channels in the grid dimensions is unnecessary, using only a downsample factor seems rather limiting. I'll change it to use the scaling factors instead, as well as finish the gradients implementation.

joaovictortr commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2373,6 +2373,14 @@ def _meshgrid(height, width):
+    grad_fn = theano.function([t_img, t_theta, t_dy], img_grad)
+
+    dy = -1 + 2 * np.random.randn(*img.shape).astype(theano.config.floatX)
+    spatialtf_grad = grad_fn(img, transform, dy)
```
Hello @notoraptor , I've implemented the initial Python part of the gradients for the spatial transformer. I couldn't get further with the C implementation, because I got stuck in this point. The gradient function compiles, however when I try to execute it here, an error is shown (``theano.gof.utils.MethodNotDefined``), which is caused by the absence of the ``perform`` method in the Op.

Do you have any ideas on what could be the problem here?

notoraptor replied:
No idea yet. Could you post the entire error message ?

joaovictortr replied:
```python
Using cuDNN version 6021 on context None
Mapped name None to device cuda0: GeForce 840M (0000:01:00.0)
E
======================================================================
ERROR: theano.gpuarray.tests.test_dnn.test_dnn_spatialtf
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jvtr/repos/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/home/jvtr/repos/Theano/theano/gof/op.py", line 901, in rval
    r = p(n, [x[0] for x in i], o, params)
  File "/home/jvtr/repos/Theano/theano/gof/op.py", line 778, in perform
    "Did you used Theano flags mode=FAST_COMPILE?"
theano.gof.utils.MethodNotDefined: ('perform', <class 'theano.gpuarray.dnn.GpuDnnTransformerGrad'>, 'GpuDnnTransformerGrad', 'Did you used Theano flags mode=FAST_COMPILE? You can use optimizer=fast_compile instead.')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jvtr/miniconda3/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/jvtr/repos/Theano/theano/gpuarray/tests/test_dnn.py", line 2382, in test_dnn_spatialtf
    spatialtf_grad = grad_fn(img, transform, dy)
  File "/home/jvtr/repos/Theano/theano/compile/function_module.py", line 898, in __call__
    storage_map=getattr(self.fn, 'storage_map', None))
  File "/home/jvtr/repos/Theano/theano/gof/link.py", line 325, in raise_with_op
    reraise(exc_type, exc_value, exc_trace)
  File "/home/jvtr/miniconda3/lib/python3.6/site-packages/six.py", line 685, in reraise
    raise value.with_traceback(tb)
  File "/home/jvtr/repos/Theano/theano/compile/function_module.py", line 884, in __call__
    self.fn() if output_subset is None else\
  File "/home/jvtr/repos/Theano/theano/gof/op.py", line 901, in rval
    r = p(n, [x[0] for x in i], o, params)
  File "/home/jvtr/repos/Theano/theano/gof/op.py", line 778, in perform
    "Did you used Theano flags mode=FAST_COMPILE?"
theano.gof.utils.MethodNotDefined: ('perform', <class 'theano.gpuarray.dnn.GpuDnnTransformerGrad'>, 'GpuDnnTransformerGrad', 'Did you used Theano flags mode=FAST_COMPILE? You can use optimizer=fast_compile instead.')
Apply node that caused the error: GpuDnnTransformerGrad{dtype='float32'}(GpuContiguous.0, GpuContiguous.0, GpuDnnTransformer{dtype='float32'}.1, Join.0, dy, _GpuDnnTransformerDescriptor{dtype='float32'}.0, Constant{1.0}, Constant{0.0})
Toposort index: 29
Inputs types: [GpuArrayType<None>(float32, 4D), GpuArrayType<None>(float32, 3D), GpuArrayType<None>(float32, 4D), TensorType(int32, vector), TensorType(float32, 4D), <theano.gof.type.CDataType object at 0x7f57b943ef98>, Scalar(float32), Scalar(float32)]
Inputs shapes: [(10, 3, 256, 256), (10, 2, 3), (10, 64, 192, 2), (4,), (10, 3, 256, 256), 'No shapes', (), ()]
Inputs strides: [(786432, 262144, 1024, 4), (24, 12, 4), (98304, 1536, 8, 4), (4,), (786432, 262144, 1024, 4), 'No strides', (), ()]
Inputs values: ['not shown', 'not shown', 'not shown', array([ 10,   3,  64, 192], dtype=int32), 'not shown', <capsule object NULL at 0x7f57b19ba450>, 1.0, 0.0]
Inputs type_num: [11, 11, 11, 5, 11, '', 11, 11]
Outputs clients: [[HostFromGpu(gpuarray)(GpuDnnTransformerGrad{dtype='float32'}.0)], [HostFromGpu(gpuarray)(GpuDnnTransformerGrad{dtype='float32'}.1)], [], [], [], []]

Backtrace when the node is created(use Theano flag traceback.limit=N to make it longer):
  File "/home/jvtr/repos/Theano/theano/gradient.py", line 1271, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/jvtr/repos/Theano/theano/gradient.py", line 966, in access_term_cache
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/jvtr/repos/Theano/theano/gradient.py", line 966, in <listcomp>
    output_grads = [access_grad_cache(var) for var in node.outputs]
  File "/home/jvtr/repos/Theano/theano/gradient.py", line 1271, in access_grad_cache
    term = access_term_cache(node)[idx]
  File "/home/jvtr/repos/Theano/theano/gradient.py", line 1107, in access_term_cache
    new_output_grads)
  File "/home/jvtr/repos/Theano/theano/gpuarray/dnn.py", line 2908, in L_op
    desc, alpha, beta)
  File "/home/jvtr/repos/Theano/theano/gof/op.py", line 615, in __call__
    node = self.make_node(*inputs, **kwargs)
  File "/home/jvtr/repos/Theano/theano/gpuarray/dnn.py", line 2928, in make_node
    context_name=context_name)()

Debugprint of the apply node: 
GpuDnnTransformerGrad{dtype='float32'}.0 [id A] <GpuArrayType<None>(float32, 4D)> ''   
 |GpuContiguous [id B] <GpuArrayType<None>(float32, 4D)> ''   
 | |GpuFromHost<None> [id C] <GpuArrayType<None>(float32, 4D)> ''   
 |   |img [id D] <TensorType(float32, 4D)>
 |GpuContiguous [id E] <GpuArrayType<None>(float32, 3D)> ''   
 | |GpuFromHost<None> [id F] <GpuArrayType<None>(float32, 3D)> ''   
 |   |theta [id G] <TensorType(float32, 3D)>
 |GpuDnnTransformer{dtype='float32'}.1 [id H] <GpuArrayType<None>(float32, 4D)> ''   
 | |GpuContiguous [id B] <GpuArrayType<None>(float32, 4D)> ''   
 | |GpuContiguous [id E] <GpuArrayType<None>(float32, 3D)> ''   
 | |Join [id I] <TensorType(int32, vector)> ''   
 | | |TensorConstant{0} [id J] <TensorType(int8, scalar)>
 | | |InplaceDimShuffle{x} [id K] <TensorType(int32, (True,))> ''   
 | | | |TensorFromScalar [id L] <TensorType(int32, scalar)> ''   
 | | |   |Cast{int32} [id M] <int32> ''   
 | | |     |ScalarFromTensor [id N] <int64> ''   
 | | |       |Shape_i{0} [id O] <TensorType(int64, scalar)> ''   
 | | |         |img [id D] <TensorType(float32, 4D)>
 | | |InplaceDimShuffle{x} [id P] <TensorType(int32, (True,))> ''   
 | | | |TensorFromScalar [id Q] <TensorType(int32, scalar)> ''   
 | | |   |Cast{int32} [id R] <int32> ''   
 | | |     |ScalarFromTensor [id S] <int64> ''   
 | | |       |Shape_i{1} [id T] <TensorType(int64, scalar)> ''   
 | | |         |img [id D] <TensorType(float32, 4D)>
 | | |InplaceDimShuffle{x} [id U] <TensorType(int32, (True,))> ''   
 | | | |TensorFromScalar [id V] <TensorType(int32, scalar)> ''   
 | | |   |Cast{int32} [id W] <int32> ''   
 | | |     |ScalarFromTensor [id X] <float64> ''   
 | | |       |Elemwise{mul,no_inplace} [id Y] <TensorType(float64, scalar)> ''   
 | | |         |TensorConstant{0.25} [id Z] <TensorType(float64, scalar)>
 | | |         |Shape_i{2} [id BA] <TensorType(int64, scalar)> ''   
 | | |           |img [id D] <TensorType(float32, 4D)>
 | | |InplaceDimShuffle{x} [id BB] <TensorType(int32, (True,))> ''   
 | |   |TensorFromScalar [id BC] <TensorType(int32, scalar)> ''   
 | |     |Cast{int32} [id BD] <int32> ''   
 | |       |ScalarFromTensor [id BE] <float64> ''   
 | |         |Elemwise{mul,no_inplace} [id BF] <TensorType(float64, scalar)> ''   
 | |           |TensorConstant{0.75} [id BG] <TensorType(float64, scalar)>
 | |           |Shape_i{3} [id BH] <TensorType(int64, scalar)> ''   
 | |             |img [id D] <TensorType(float32, 4D)>
 | |_GpuDnnTransformerDescriptor{dtype='float32'} [id BI] <CDataType{cudnnSpatialTransformerDescriptor_t}> ''   
 | | |Cast{int32} [id M] <int32> ''   
 | | |Cast{int32} [id R] <int32> ''   
 | | |Cast{int32} [id W] <int32> ''   
 | | |Cast{int32} [id BD] <int32> ''   
 | |Constant{1.0} [id BJ] <float32>
 | |Constant{0.0} [id BK] <float32>
 |Join [id I] <TensorType(int32, vector)> ''   
 |dy [id BL] <TensorType(float32, 4D)>
 |_GpuDnnTransformerDescriptor{dtype='float32'} [id BI] <CDataType{cudnnSpatialTransformerDescriptor_t}> ''   
 |Constant{1.0} [id BJ] <float32>
 |Constant{0.0} [id BK] <float32>
GpuDnnTransformerGrad{dtype='float32'}.1 [id A] <GpuArrayType<None>(float32, 3D)> ''   
GpuDnnTransformerGrad{dtype='float32'}.2 [id A] <NullType> ''   
GpuDnnTransformerGrad{dtype='float32'}.3 [id A] <NullType> ''   
GpuDnnTransformerGrad{dtype='float32'}.4 [id A] <NullType> ''   
GpuDnnTransformerGrad{dtype='float32'}.5 [id A] <NullType> ''   

Storage map footprint:
 - img, Input, Shape: (10, 3, 256, 256), ElemSize: 4 Byte(s), TotalSize: 7864320 Byte(s)
 - dy, Input, Shape: (10, 3, 256, 256), ElemSize: 4 Byte(s), TotalSize: 7864320 Byte(s)
 - GpuContiguous.0, Shape: (10, 3, 256, 256), ElemSize: 4 Byte(s), TotalSize: 7864320 Byte(s)
 - GpuDnnTransformer{dtype='float32'}.out, Shape: (10, 3, 64, 192), ElemSize: 4 Byte(s), TotalSize: 1474560 Byte(s)
 - GpuDnnTransformer{dtype='float32'}.1, Shape: (10, 64, 192, 2), ElemSize: 4 Byte(s), TotalSize: 983040 Byte(s)
 - theta, Input, Shape: (10, 2, 3), ElemSize: 4 Byte(s), TotalSize: 240 Byte(s)
 - GpuContiguous.0, Shape: (10, 2, 3), ElemSize: 4 Byte(s), TotalSize: 240 Byte(s)
 - Join.0, Shape: (4,), ElemSize: 4 Byte(s), TotalSize: 16 Byte(s)
 - TensorConstant{0.75}, Shape: (), ElemSize: 8 Byte(s), TotalSize: 8.0 Byte(s)
 - TensorConstant{0.25}, Shape: (), ElemSize: 8 Byte(s), TotalSize: 8.0 Byte(s)
 - Constant{1.0}, Shape: (), ElemSize: 4 Byte(s), TotalSize: 4.0 Byte(s)
 - Constant{0.0}, Shape: (), ElemSize: 4 Byte(s), TotalSize: 4.0 Byte(s)
 - TensorConstant{0}, Shape: (), ElemSize: 1 Byte(s), TotalSize: 1.0 Byte(s)
 TotalSize: 26051081.0 Byte(s) 0.024 GB
 TotalSize inputs: 15728905.0 Byte(s) 0.015 GB



----------------------------------------------------------------------
Ran 1 test in 2.825s

FAILED (errors=1)
```

notoraptor replied:
Strange.

But, to start, please apply my comments about `grad_undefined()`. Currently, `make_node()` returns 6 outputs, but C function (`dnn_sptf_grad()`) will generate only 2 outputs. `make_node()` must not return any `grad_undefined()`.

Let's first stabilize the code, and then we will check errors.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2800,6 +2800,204 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+                   theano.gradient.grad_undefined(self, 2, grid_dims),
+                   theano.gradient.grad_undefined(self, 3, desc),
+                   theano.gradient.grad_undefined(self, 4, alpha),
+                   theano.gradient.grad_undefined(self, 5, beta)]
```
It is not the role of `make_node()` to return some `grad_undefined()`s, this should be do in `L_op()` or `grad()` only.

You should better copy the strategy we follow for convolution: there are 2 gradient ops: GpuDnnConvGradI (gradient w/ inputs) and GpuDnnConvGradW (gradient w/ weights). Here gradient w/ inputs should be w/ img, and gradient w/ weights should be w/ theta, I think. See here: https://github.com/Theano/Theano/blob/master/theano/gpuarray/dnn.py#L580 .

notoraptor replied:
I think at the end you may have 4 classes:
- GpuDnnTransformerDescriptor
- GpuDnnTransformer
- GpuDnnTransformerGradI
- GpuDnnTransformerGradW

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2800,6 +2800,204 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
     return [rval]
 
 
+class _GpuDnnTransformerDescriptor(COp):
```
Remove the underscore at top of name: `_GpuDnnTransformerDescriptor` -> `GpuDnnTransformerDescriptor`.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2800,6 +2800,204 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        DnnBase.__init__(self, ["c_code/dnn_sptf.c"], "dnn_sptf")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, grid_dims, desc, alpha=None, beta=None):
```
`alpha` and mostly `beta` seems useless if an output is not given.

More generally, see how cuDNN convolutions ops (GpuDnnConvDesc, GpuDnnConv, GpuDnnConvGradI, GpuDnnConvGradW) are implemented and used to get an idea about how to implement cuDNN spatial transformations ops.

Also, in cuDNN doc, I remark that there are functions for grid generation forward/backward, and then functions for grid sampling forward/backward. You initially separated grid generation and grid sampling, that may allow to apply gradient for each op. Now you seem to merge both operations into just one. Are you sure an user won't need to work with these two steps (generation and sampling) separately ?

joaovictortr replied:
Most implementations in Theano use hardcoded values for ``alpha`` and ``beta`` (e.g. ``dnn_softmax``), do you think it is also the case here?

I don't think the user will want (or need) the grid of coordinates - generated by the forward grid, since it is only used by the sampler to map the pixels that must be sampled from the input images. Also, the only outputs the user is interested from the spatial transformer - in the forward pass, are the transformed images.

In the case of gradients, this is somewhat different, since the user will also want the gradients of the affine transformations - besides the gradients of the inputs, so it is possible to backpropagate them to the localization network. Therefore, there are separate classes for each of these gradients, which generated by separate backward functions of cuDNN. However, they are returned together in the ``L_op`` method of ``GpuDnnTransformer``, because the output of the backward function of the sampler is the input of the backward function of the grid generator.

notoraptor replied:
Ok, so you can keep `GpuDnnTransformer` as 1 op.


For alpha and beta, there are 2 choices:
- either you keep it as op inputs, and then you should also provide an `out` input (as GpuDnnConv does, for example), so that the user can get `out = alpha*computed + beta*out`
- or you hardcode these values (alpha = 1 and beta = 0), and in such case you remove it from op inputs.

The better thing would be to keep it as op inputs (as GpuDnnConv), to have the code ready for any future extension / improvement.

notoraptor commented:
Hello, @joaovictortr I made some general comments (not yet reviewed the whole code).

joaovictortr commented:
Hello @notoraptor , do you know how can I verify the gradients of the spatial transformer? I've tried doing it separately with each gradient class, using ``utt.verify_grad``, but to no avail. Checking directly with ``L_op`` from ``GpuDnnTransformer`` returns more than one output, and so it cannot be checked with ``verify_grad``, at the moment.

I think this is the only part missing to complete the implementation (before any reviews).

notoraptor commented:
@joaovictortr could you try to have 2 verify_grad, one per grad (ie. one for input and one for theta) ?

joaovictortr commented:
@notoraptor using two separate verify_grad functions would require a manual setup of the gradient's ops. I'm confused, because if I setup the gradient Op manually, and then run verify_grad on it, it would be computing the gradient of the gradient Op, which is currently not defined. Could you elaborate on it?

notoraptor commented:
@joaovictortr I have not tested it, but GpuDnnTransformer has currently a L_op() that will compute GpuDnnTransformerGradI and GpuDnnTransformerGradT, right ?

So, does something like this work ?
```python
# Functor to verify grad wrt/ input
def functor_wrt_i(input):
	transformed_input = GpuDnnTransformer(...)(input, theta, ...)
	grad = T.grad(T.mean(transformed_input), input)
	return grad

# Functor to verify grad wrt/ theta
def functor_wrt_t(theta):
	transformed_input = GpuDnnTransformer(...)(input, theta, ...)
	grad = T.grad(T.mean(transformed_input), theta)
	return grad

# Then:
utt.verify_grad(functor_wrt_i, [some_input])
utt.verify_grad(functor_wrt_t, [some_theta])
```
?

notoraptor commented:
PS: with `T.mean` or any other relevant op that returns a scalar.

joaovictortr commented:
@notoraptor using the gradient wrt to a cost, invokes the gradient of both ``GpuDnnTransformerGradI`` and ``GpuDnnTransformerGradT``. I have tried providing the implementations of ``grad`` to both operations, but I have not yet succeeded.

What is not clear to me is how the gradients of the transformed images will be retrieved in the ``grad`` or ``L_op`` function. In the case of ``grad`` or ``L_op``, ``ouput_gradients`` should contain the gradient of the cost function with respect to the images (input) or the transformation tensor, correct?

notoraptor commented:
Hello @joaovictortr , about gradients, I had provided a full explanation in CTC PR : https://github.com/Theano/Theano/pull/5949#discussion_r120652640 !

`output_gradients` will contain the gradient of the Theano's graph output (in wich GpuDnnTransformer is used) with respect to each Op's output (`dC/dy`). So, you have to compute `dy/dx` for each Op's input `x`, and multiply `dC/dy` by `dy/dx` to get `dC/dx` that is the gradient of the Theano's graph ouput with respect to a particular Op input `x`.

I will made a full review in next minutes, as it seems gradients are the last things to implement.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def make_node(self, img, theta, grid, grid_dims, dy, desc, alpha, beta):
+        context_name = infer_context_name(img)
+
+        if img.ndim != 4:
```
Check if after applying `as_gpuarray_variable`, not before.

joaovictortr replied:
Done.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    """
+
+    # inp is a 4D tensor with shape: (num_inputs, num_channels, height, width)
+    assert img.ndim == 4
```
At this point you are not sure that img has a `ndim` field. You should apply `as_gpuarray_variable()` before.

joaovictortr replied:
Fixed as well.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        return ['cudnn.h', 'cudnn_helper.h']
+
+    def c_header_dirs(self):
+        return [os.path.dirname(__file__), config.dnn.include_path]
```
Done.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        return ['cudnn']
+
+    def c_lib_dirs(self):
+        return [config.dnn.library_path]
```
Done.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        assert cudnn.cudnnDataType_t.has_alias(dtype)
+        self.dtype = dtype
+
+    def make_node(self, dimensions):
```
I've used it as a Theano variable, otherwise the grid shape would have to passed by the user, which would repeat the need to pass information on the number of images and channels. Furthermore, the grid dimensions tuple is setup according to the shape of the input image's shape, and the image might be a Theano variable as well.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        DnnBase.__init__(self, ["c_code/dnn_sptf.c"], "APPLY_SPECIFIC(dnn_sptf)")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, output, grid_dims, desc, alpha=None, beta=None):
```
Currently, I can also retrieve the grid dimensions by taking the inputs from the owner of ``desc``.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        self.dtype = dtype
+
+    def make_node(self, img, theta, output, grid_dims, desc, alpha=None, beta=None):
+        assert theta.dtype in ('float16', 'float32', 'float64')
```
Done.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,263 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        output = as_gpuarray_variable(output, context_name)
+
+        grid = GpuArrayType(dtype=self.dtype,
+                            broadcastable=img.type.ndim * (False,),
```
Done.

joaovictortr commented:
Hello @notoraptor, thanks for the reminder and for pointing out to the full explanation of the gradients! I have also checked the documentation on gradients again, as well. I've also addressed your comments, and secured the inputs before any checks. The problem I ran into now, is the conversion of an object to a gpuarray.

For now, the grid dimensions are being passed as a Theano variable, otherwise I think the user would have to provide the shape of the grid, and would have to specify the number of images and channels again, in the call to the spatial transformer. The dimensions can be retrieved by taking the first input of the descriptor node, which is the tuple containing the dimensions.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    operations.
+    """
+
+    __props__ = ('dtype',)
```
Remark that dtype is not used at all in `make_node()`.

In fact, I suggest you totally remove `dtype` prop and param from all ops related to spatial transformation. Output dtype should simply be the same as input (img) dtype.

joaovictortr replied:
I've removed dtype from all of them, except the descriptor, following the design of ``GpuDnnConv``, as the cuDNN datatype (``cudnnDataType_t``) is required in the descriptor setup.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gi.c"], "APPLY_SPECIFIC(dnn_sptf_gi)")
+        self.dtype = dtype
+
+    def make_node(self, img, theta, grid, dy, desc, alpha, beta):
```
Whenever alpha and beta are used, there must be an `output` in inputs list. Recall how alpha and beta are used:
```python
final_output = alpha * computed_output + beta * given_output
```
where `given_output` is the output given in `make_node()`.

**NB**: As it seems too complicated, I ultimately suggest you remove alpha and beta support. I suggested to support it to facilitate any future evolution of code, but in fact they are not currently used, event for cuDNN convolutions ops (at least in public interface). Sorry for that new request for interface change, but I think it will make your work easier.

So, remove alpha and beta support for transformer op and transformer grad ops. Just use `alpha=1` and `beta=0` in C code.

joaovictortr replied:
Indeed, it makes the work easier, so that I can get to a stable base, with the main problem being the gradients now. It can be added later on, if you think it is worth it.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,233 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    Spatial Transformer implementation using Theano from Lasagne
+    Original author: skaae (https://github.com/skaae)
+    """
+    def spatialtf_cpu(theta, inp, scale_height, scale_width, border_mode='nearest'):
```
It seems you have finally implemented a CPU version for spatial transformation ! Could you move it to CPU part of theano (somewhere relevant in `theano.tensor`) ? Maybe a right place should be `theano/tensor/spatialtf.py`, as this could be used even for non-NNET operations.

If so, add a consistent docstring (infos about args) and a doc entry. We will discuss about it later, but this seems close to be a third part for your GSoC !

PS: I have not yet reviewed the whole code.

joaovictortr replied:
That is correct, the CPU implementation was taken mainly from Lasagne's spatial transformer. I also think it can be ported to Theano, and it would be a good replacement for the third part of GSoC.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        self.dtype = dtype
+
+    def make_node(self, dimensions):
+        dimensions = as_tensor_variable(dimensions)
```
You haven't checked `dimensions` dtype, but in C code you assume that it's an integer array (as you use `(int*) PyArray_GETPTR1(...)`).

What I suggest is:
- check that `dimensions` ndim is 1 (we expect a vector).
- check that `dimensions` dtype is an integer type (dtype should be in `theano.tensor.basic.int_dtypes`).
- apply as cast to force dimensions to have a specific type in C code (this will let the user use any integer type he wants). I suggest you use int64 as it is the biggest supported integer type. Example: `dimensions = theano.tensor.basic.cast(dimensions, 'int64')`
- Then in C code use `(npy_int64*)` instead of `(int*)`. Also, don't forget that an `int` in C code does not have a standard size accross all systems and all compilers, so be careful when you can something to an int !

joaovictortr replied:
You'are absolutely correct, I'll follow your suggestions, of course. I've used int types, because that is how the interface of cuDNN is specified, but it is essential to check the Python object, before reading from it in the C code.

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_desc.c:
```diff
@@ -0,0 +1,46 @@
+{
+    cudnnStatus_t err;
+
+    const int nimages = *((int *) PyArray_GETPTR1(dims, 0));
```
See my comments below (in `dnn.py`), but it is quite dangerous to cast something to an int without some checkings.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+            raise TypeError('img type must be floating-point')
+
+        theta = gpu_contiguous(as_gpuarray_variable(theta, context_name))
+        assert theta.dtype in ('float16', 'float32', 'float64')
```
Done.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        grid_dims = as_tensor_variable(desc.owner.inputs[0])
+
+        dy = as_gpuarray_variable(dy, context_name)
+        if img.ndim != 4:
```
Fixed.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+        dy = as_gpuarray_variable(dy, context_name)
+        if img.ndim != 4:
+            raise TypeError('img must have 4 dimensions.')
```
Fixed.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,295 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+                            context_name=context_name)()
+        dgrid = GpuArrayType(dtype=self.dtype,
+                             broadcastable=img.type.ndim * (False,),
+                             context_name=context_name)()
```
Done.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        self.precision = precision
+
+    def make_node(self, dimensions):
+        dimensions = as_tensor_variable(dimensions)
```
After this line, check that dimension.ndim is 1 (as a vector is expected).

Also, you must check dimension dtype (integer types) and cast it (-> int64) here also, as you did in GpuDnnTransformer, because you read it in C code.

notoraptor replied:
Don't forget this (check dimensions ndim == 1, check dimensions dtype in integer_dtypes, and cast it to int64).

notoraptor commented on theano/gpuarray/c_code/dnn_sptf.c:
```diff
@@ -0,0 +1,203 @@
+    }
+
+    // Obtain grid dimensions
+    num_images = (int) *( (npy_int *) PyArray_GETPTR1( grid_dims, 0 ) );
```
As you have casted grid_dims to int64 in Python code, it should be `npy_int64` here and in next lines, not `npy_int`.

notoraptor replied:
Don't forget this (npy_int -> npy_int64).

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_gi.c:
```diff
@@ -0,0 +1,172 @@
+
+int
+APPLY_SPECIFIC(dnn_sptf_gi)(PyGpuArrayObject * input,
+                            PyGpuArrayObject * theta,
```
It seems you don't need theta in this code. If so, it must be removed from here and from Op's inputs in Python code.

notoraptor replied:
Finally, you don't need theta at all in this file, so remove it totally.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def L_op(self, inputs, outputs, grads):
+        img, theta, grid_dims, desc = inputs
+        _, grid = outputs
+        dy = grads[0]
```
Hello @notoraptor , the issue I've mentioned, of converting an object to a ``GpuArrayVariable`` is caused here, because dy is a ``TensorVariable``. ``dy`` is passed to ``L_op`` through the ``verify_grad`` function. Is there a way to convert a ``TensorVariable`` to a ``GpuArrayVariable``?

notoraptor replied:
Just apply `as_gpuarray_variable(dy)`, it should work.

joaovictortr replied:
Using ``as_gpuarray_variable``, gives the exact same [error in the gradient test](https://gist.github.com/joaovictortr/cf4e59ffc8a8a8c130f128e0513aa44f)

joaovictortr replied:
Furthermore, the ``dy`` variable is of type ``DisconnectedType``, that's why it fails with both ``as_gpuarray_variable`` and ``as_tensor_variable``.

notoraptor replied:
Strange that it's a Disconnected.

I will do more investigations tomorrow. For now, make sure that anything works except the gradients.

Also, re-read all your `make_node()` methods and make sure that every input is checked correctly (dtype, ndim, as_gpuarray/gpu_contiguous/cast if necessary, etc.). For every op, including gradient Ops !

notoraptor commented on theano/gpuarray/c_code/dnn_sptf.c:
```diff
@@ -0,0 +1,203 @@
+    if ( PyGpuArray_DIM( input, 0 ) != num_images )
+    {
+        PyErr_Format( PyExc_RuntimeError,
+            "GpuDnnTransformer: expected input to have %d inputs, got %d inputs.",
```
Consider rewrite this message, e.g.: "GpuDnnTransformer: expected batch size %d, got %d." .

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_gt.c:
```diff
@@ -0,0 +1,43 @@
+
+    const size_t dtheta_dims[3] = { num_images, 2, 3 };
+
+    if ( theano_prep_output( dtheta, 3, &(dtheta_dims[0]), dgrid->ga.typecode,
```
`&(dtheta_dims[0])` ->just `dtheta_dims`

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    Gradients of inputs of the spatial transformer
+    """
+    __props__ = ()
+    _cop_num_inputs = 5
```
Should be 4 now (as you don't need theta here).

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    def __init__(self, dtype=theano.config.floatX):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gi.c"], "APPLY_SPECIFIC(dnn_sptf_gi)")
+
+    def make_node(self, img, theta, grid, dy, desc):
```
theta is useless for this op, so remove it.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+        return Apply(self, inputs, outputs)
+
+    def L_op(self, inputs, outputs, grads):
+        img, theta, grid, dy, desc = inputs
```
theta won't be in inputs anymore.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    def connection_pattern(self, node):
+        # not connected to desc
+        return [[1, 1], [1, 1], [1, 1], [1, 1], [0, 0]]
```
Update this too, as theta is removed.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,273 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+
+    def make_node(self, dgrid, desc):
+        context_name = infer_context_name(dgrid)
+        dtheta = GpuArrayType(dtype=dgrid.dtype,
```
Before this line:
- apply `as_gpuarray_variable` to dgrid.
- check dgrid dtype (in float16, float32 or float64)
- check dgrid ndim (expected 4).

joaovictortr commented:
Hello @notoraptor , I'll follow your design, as I have extensively tested different approaches and had the same issues with ``DisconnectedType``s appearing on the ``L_op`` methods, which you have also described. I would add the CPU spatial transformer from Lasagne (separate PR?), and the optimization to lift the op to the GPU.

notoraptor commented:
Hello @joaovictortr ! About the CPU part, let's do it in another PR, to avoid this PR grow too much (it already takes about 1000 lines of code) !

Let's first finish with GPU part here.

joaovictortr commented:
@notoraptor I've split the implementations and addressed your comments as well. The gradients seem to work now, although it takes quite some time to compute them. The gradient tests are failing because of the difference between the numeric and analytic values, but there are no more problems related to ``DisconnectedType`` objects.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+                for node in grad_fn.maker.fgraph.toposort()])
+
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradT)
+                for node in grad_fn.maker.fgraph.toposort()])
```
I think all this code is too complicated if the goal is just to check if generated function contains the expected ops !
Also:
1) for gradient test, we don't need to mimic the usual images format, we can just directly generate an NCHW image.
2) To iterate over function graph nodes, you can use `fn.maker.fgraph.apply_nodes` instead of `.toposort()`, it will be faster.

Why not just something like that:
```python
    theta = T.tensor3()
    inputs = T.tensor4()
    output = dnn.dnn_spatialtf(inputs, theta, 0.25, 0.75)
    output_mean = T.mean(output)
    mean_gi = T.grad(output_mean, [inputs])
    mean_gt = T.grad(output_mean, [theta])
    f_gi = theano.function([inputs, theta], mean_gi)
    f_gt = theano.function([inputs, theta], mean_gt)
    assert any(isinstance(node.op, dnn.GpuDnnTransformerGradI) for node in f_gi.maker.fgraph.apply_nodes)
    assert any(isinstance(node.op, dnn.GpuDnnTransformerGradT) for node in f_gt.maker.fgraph.apply_nodes)
    inputs_shape = (1, 3, 10, 10)
    inputs_val = np.random.random(inputs_shape).astype(theano.config.floatX)
    theta_val = np.random.random((inputs_shape[0], 2, 3)).astype(theano.config.floatX)
    # Just check that everything computes.
    f_gi(inputs_val, theta_val)
    f_gt(inputs_val, theta_val)```

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        op = dnn.dnn_spatialtf(img, theta)
+        return op
+
+    utt.verify_grad(fn_wrt_i, [img, theta], mode=mode_with_gpu)
```
Also, I have no gradient error with a theta generated randomly.
```python
theta_val = np.random.random((inputs_shape[0], 2, 3)).astype(theano.config.floatX)
```

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    utt.seed_rng()
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values in [0, 255])
+    img_dims = (1, 256, 256, 3)  # images are usually NHWC
```
It seems gradient tests take time because inputs are quuite big. Use smaller images (e.g. width 10 and height 10).

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,289 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+    _f16_ok = True
+
```
Also add `check_input = False` here. It will help Theano to not generate two many variants of C code, as we either force types (e.g. with int64) or don't use any macro related to inputs or outputs dtypes into our C codes.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,289 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    _cop_num_inputs = 3
+    _cop_num_outputs = 1
+    _f16_ok = True
+
```
Also add `check_input = False` here.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,289 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    _cop_num_inputs = 4
+    _cop_num_outputs = 2
+    _f16_ok = True
+
```
Also add `check_input = False` here.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2833,6 +2833,289 @@ def local_abstractconv3d_cudnn_graph(op, context_name, inputs, outputs):
+    _cop_num_inputs = 2
+    _cop_num_outputs = 1
+    _f16_ok = True
+
```
Also add `check_input = False` here.

joaovictortr commented:
Thank you very much for the thorough review @notoraptor ! I've addressed your comments in the latest commits.

Do you think the docstrings should be further extended with information about the tensor format, for example?

notoraptor commented:
Yes, extend docstrings with infos about `make_node()` inputs everywhere it's relevant.

NB: About the last point (`Create tutorial/example to showcase how to use Spatial Transformer Networks in Theano`), maybe this could be done in the 3rd PR with CPU implementation.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2738,305 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+    out_dims = (img.shape[0], img.shape[1],
+                img.shape[2] * scale_height,
+                img.shape[3] * scale_width)
+    out_dims = tuple([as_scalar(v).astype('int64') for v in out_dims])
```
Instead of applying `as_scalar()` and then creating a tuple, just use:
```python
out_dims = np.asarray(out_dims, dtype='int64')
```

joaovictortr replied:
I've kept it this way because when img is a Variable, trying to create a NumPy array from ``out_dims`` fails, because the shape of the Variable is not known in advance.

notoraptor replied:
Ah ok, you're right.

So, don't forget to apply `T.ceil` !

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,214 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+        # grid of (x_t, y_t, 1), eq (1) in ref [1]
+        out_height = T.cast(T.ceil(height * scale_height), 'int64')
+        out_width = T.cast(T.ceil(width * scale_width), 'int64')
```
I just noticed that you call `ceil` here. But in `dnn_spatialtf()`, when you do `as_scalar(v).astype('int64')`, it's a floor that is done. That could be a problem, isnt'it ?

We should choose 1 rounding strategy and use it in both CPU and GPU functions. Maybe you should use ceil in both places. In such case, my comment about it in `dnn_spatialtf` will become:
```python
out_dims = np.ceil(out_dims).astype('int64')
```

joaovictortr replied:
Indeed, I hadn't notice that in the CPU implementation. I'll use ceil, as you've suggested.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,214 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        height_f = T.cast(height, theano.config.floatX)
+        width_f = T.cast(width, theano.config.floatX)
+
+        # scale coordinates from [-1, 1] to [0, width/height - 1]
```
`width/height` is a little confusing. Write `dimension` instead.

joaovictortr replied:
I've changed to dimension, as requested.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,214 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        grid = T.concatenate([x_t_flat, y_t_flat, ones], axis=0)
+        return grid
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values
```
You don't use 256x256 resolution currently!

joaovictortr replied:
Fixed.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,214 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    assert any([isinstance(node.op, dnn.GpuDnnTransformerGradT)
+                for node in f_gt.maker.fgraph.apply_nodes])
+
+    # Generate random set of RGB images with 256x256 resolution (pixel values
```
You don't use 256x256 resolution currently!

joaovictortr replied:
Fixed.

joaovictortr commented:
Hello @notoraptor , I've addressed your comments in the latest commits. I've also added the docstrings for ``GpuDnnTransformerGrid`` and ``GpuDnnTransformerSampler`` Ops.

Do you think I can move onto the CPU implementation?

notoraptor commented:
jenkins test this

notoraptor commented:
Still errors with verify grad (expected vs value).

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,215 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        out = dnn.dnn_spatialtf(inputs, theta)
+        return out
+
+    utt.verify_grad(grad_functor, [inputs_val, theta_val], mode=mode_with_gpu)
```
Decrease absolute tolerance to 1e-3 (currently 1e-4) seems to resolve the problem. Can you check it on your computer ? Does this test pass with default absolute tolerance on your computer ?

```python
utt.verify_grad(grad_functor, [inputs_val, theta_val], mode=mode_with_gpu, abs_tol=1e-3)
```

NB: Current spatialtf tests use floatX, which can be float16, float32 or float64, and appropriate tolerance may change from a float type to another (especially with float16). Check on your computer if tests pass with every possible value for floatX.

joaovictortr replied:
Indeed, it solves the problem when floatX is float64. However, i'm getting absolute and relative errors of 0.12 and 0.1, respectively, when using float32. Could you confirm that, please? I also have access to some Titan cards, I'll check on those as well.

Testing with float16 fails completely with the absolute error not being finite. I think float16 should be disabled if this is confirmed in other GPUs, what do you think?

The relevant GPU specs for my machine are:
GPU model: Geforce 840M with compute capability of 5.0 (Maxwell)
CUDA version: 8
cuDNN version: 6.0.21

joaovictortr replied:
At least, warn the user of these precision problems that might occur, when he wants to compute the gradients.

joaovictortr replied:
@notoraptor , gradient computations fail with float16, at least on my machine.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,215 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    st_cpu_func = theano.function([t_img, t_theta], st_cpu, mode=mode_without_gpu)
+    img_out_cpu = st_cpu_func(img, theta)
+
+    utt.assert_allclose(img_out_cpu, img_out_gpu, rtol=1e-2, atol=1e-2)
```
Is it expected that we need such low atol (1e-2) ? Does it fail with 1e-3, 1e-4, 1e-5 ?

joaovictortr replied:
It actually passes now with lower tolerance values. In my machine, it works with a tolerance of 1e-5 using float32.

notoraptor replied:
Ok (PS: I meant "such high atol", not "such low atol"). Check if it works with floatX=float16 too. Let's try to use a lower atol.

If a lower atol is not convenient for float16, use a specific atol for float16.

Check gradient test also.

joaovictortr replied:
It works, however with a much higher atol (1e-1) and rtol (1e-2). Since these differences are negligible when working with images, do you think an exception should be made for float16?

notoraptor replied:
For float16, try with rtol=6e-2 and atol with its default value. That's what we use in some tests for cuDNN convolutions: https://github.com/Theano/Theano/blob/master/theano/gpuarray/tests/test_dnn.py#L1180

notoraptor commented:
Ok @joaovictortr with your last commits:

`dnn_spatialtf` passes with float64 and float32, fails with float16, but with the following error:

```
-------------------- >> begin captured logging << --------------------
theano.gof.opt: ERROR: SeqOptimizer apply <theano.gof.opt.EquilibriumOptimizer object at 0x000000000656BE80>
theano.gof.opt: ERROR: Traceback:
theano.gof.opt: ERROR: Traceback (most recent call last):
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\gof\opt.py", line 241, in apply
    sub_prof = optimizer.optimize(fgraph)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\gof\opt.py", line 87, in optimize
    ret = self.apply(fgraph, *args, **kwargs)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\gof\opt.py", line 2475, in apply
    lopt_change = self.process_node(fgraph, node, lopt)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\gof\opt.py", line 1995, in process_node
    replacements = lopt.transform(node)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\tensor\blas.py", line 1721, in local_dot22_to_ger_or_gemv
    rval = ger(zeros, one, xv, yv)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\gof\op.py", line 615, in __call__
    node = self.make_node(*inputs, **kwargs)
  File "C:\Users\notoraptor\mila\dev\git\theano\theano\tensor\blas.py", line 335, in make_node
    raise TypeError('only float and complex types supported', x.dtype)
TypeError: ('only float and complex types supported', 'float16')
```

Can you check it ?

`dnn_spatialtf_grad` passes with float64, but fails with float32 like on your machine (`val1 = 0.514990      ,  val2 = 0.636695`). It also fails with float16 with a `abs_err not finite` message.

I am testing on a GeForce GTX 1050.

As you suggested, we can warn the user when he wants to compute gradient with precision float32 or float16.

joaovictortr commented:
Hello @notoraptor , the error is also related to the tolerances, there are some cases where a higher tolerance is necessary. For float16, I'm getting absolute and relative errors of up to 0.12 and 0.001, respectively. Raising the absolute tolerance to 1e-0 seems to solve the issue for float16.

Regarding the optimization error, I'm looking into what may be causing it, as it only happens when using float16.

notoraptor commented:
1e0 (ie. just 1) is too high for tolerance.

I'll check that tomorrow again, but maybe should we disable float16 and support only float32 and float64.

NB: For float16 with `test_dnn_spatialtf`, check also my comment here: https://github.com/Theano/Theano/pull/6061#discussion_r129954114

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        out = dnn.dnn_spatialtf(inputs, theta)
+        return out
+
+    atol, rtol = 1e-3, 1e-3
```
Let them to None by default.

joaovictortr replied:
Fixed.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    # Generate random set of RGB images(pixel values in [0, 255]). The set of
+    # images is generated in the expected (NCHW) format
+    input_dims = (5, 3, 16, 16)
+    inputs_val = np.random.randint(low=0, high=256, size=input_dims).astype(theano.config.floatX)
```
Here also, using integers may be the source of problems as values are too high. For gradient test, I suggest you use random floating values in [0;1] (`np.random.random`).

joaovictortr replied:
I've tested with the inputs in range [0, 1], and they improve considerably the error, even on float16. However, now the error on float16 is finite, but still high (order of 1).

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    input_dims = (5, 3, 16, 16)
+    inputs_val = np.random.randint(low=0, high=256, size=input_dims).astype(theano.config.floatX)
+    # Tensor with transformations
+    theta_val = np.random.random((input_dims[0], 2, 3)).astype(theano.config.floatX)
```
Check how atol varies when you downgrade theta and inputs_val. e.g.:
```python
inputs_val /= 10 # or even /= 100
theta_val /= 10 # or even /= 100
```

joaovictortr replied:
It drops one or two orders of magnitude, depending on the inputs and theta. However, reducing further from 1000 has shown no benefits at all. The gradients in float16 still fail, but the error is now in the order of 1 or 10 (which is still too high).

joaovictortr commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+    atol, rtol = None, None  # use default absolute and relative error tolerances
+    if theano.config.floatX == 'float16':  # float16 requires higher tolerances
+        atol, rtol = 1e-0, 1e-2
```
Indeed, it is way too high, and I think you are correct. Yes, I have used integers to mimic the pixel values from images.

Using inputs to the range [0,1] improves considerably the accuracy of the results.

Rounding usually makes the test fail, because there cases in which it generates a difference of 1 between pixels of the generated images, but this difference wouldn't be noticeable in the output (if we're using images).

joaovictortr commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+    atol, rtol = 1e-3, 1e-3
+    if theano.config.floatX == 'float32':  # use higher error tolerances with float32
+        atol, rtol = 1e-0, 1e-1
```
I've done that, and because it is using inputs in the range [0,1], the tolerance of 5e-2 is now sufficient, even on float16.

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2738,342 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gi.c"], "APPLY_SPECIFIC(dnn_sptf_gi)")
+        if theano.config.floatX == 'float16' or theano.config.floatX == 'float32':
```
I've removed the warnings, there is just a problem with the gradients on float16 not being under the tolerance (order of 1 or 10).

joaovictortr commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2738,342 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+
+    def __init__(self):
+        DnnBase.__init__(self, ["c_code/dnn_sptf_gt.c"], "APPLY_SPECIFIC(dnn_sptf_gt)")
+        if theano.config.floatX == 'float16' or theano.config.floatX == 'float32':
```
Removed.

joaovictortr commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,225 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+    # Generate random set of RGB images (pixel values in [0, 255]). The set of
+    # images is generated in the expected (NCHW) format
+    img_dims = (5, 3, 16, 16)
+    img = np.random.randint(low=0, high=256, size=img_dims).astype(theano.config.floatX)
```
Using the values in the range [0; 1] does indeed solve the problem, except for float16 in the gradients' test.

joaovictortr commented:
Hello @notoraptor , I've applied your suggestions, and they indeed solve most of the problems with different precisions. Changing the range of inputs from [0; 255] to [0; 1] does indeed improves considerably the accuracy, and the test now passes with both float32 and float16 with lower error tolerances.

I've tried reducing the values of theta and inputs in the gradients test, however they still fail with float16, with an order of magnitude of 1 or 10 in the error values.

notoraptor commented:
Ok. So now, only gradient test fails and only with float16, right ?

notoraptor commented:
ok to test

joaovictortr commented:
Exactly, it only fails on the gradients' test, with float16

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+    # Tensor with transformations
+    theta_val = np.random.random((input_dims[0], 2, 3)).astype(theano.config.floatX)
+    theta /= 100
```
Add a comment to explain why you divide theta here (like here: https://github.com/Theano/Theano/blob/master/theano/gpuarray/tests/test_dnn.py#L1087 ).

joaovictortr replied:
Done.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+        return out
+
+    atol, rtol = None, None
+    if theano.config.floatX == 'float16' or theano.config.floatX == 'float32':
```
Ok. Then choose a specific rtol for float32 and a specific rtol for float16 too (don't use the same rtol for both), so that tests pass for all precisions.

joaovictortr replied:
Done.

notoraptor commented on theano/gpuarray/tests/test_dnn.py:
```diff
@@ -2298,3 +2298,221 @@ class Cudnn_grouped_conv(Grouped_conv_noOptim):
+
+    atol, rtol = None, None
+    if theano.config.floatX == 'float16' or theano.config.floatX == 'float32':
+        atol, rtol = 5e-2, 5e-2
```
Is it necessary to set both tolerances ? If not, try to set only one of them for each precision (float16 and float32).

joaovictortr replied:
No, only the relative tolerance, I've fixed that. I've left the tolerance for float16 is way too high, but otherwise the test fails.

notoraptor commented:
Flake8 !

/home/jenkins/workspace/Theano_PR/theano/gpuarray/tests/test_dnn.py:2505:64: W291 trailing whitespace

abergeron commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2738,336 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+    def __init__(self, precision=theano.config.floatX):
+        COp.__init__(self, ["c_code/dnn_sptf_desc.c"], "APPLY_SPECIFIC(dnn_sptf_desc)")
+        assert cudnn.cudnnDataType_t.has_alias(precision)
+        self.precision = precision
```
Please mirror what the pooling code does and just create a descriptor in each op, setting the appropriate values at runtime.

We are trying to avoid expanding on the number of "create a descriptor" ops.  This will also simplify the code for the ops below.

notoraptor replied:
Ok.
@joaovictortr, so, I suggest this new scheme, in which we have no `GpuDnnTransformerDesc` anymore:
```python
# Recall: out_dims can be inferred as long as we have img and grid:
out_dims
	# out_dims = N,C,H,W:
	# N: img.shape[0], grid.shape[0]
	# C: img.shape[1]
	# H: grid.shape[1]
	# W: grid.shape[2]

GpuDnnTransformerGrid(theta, out_dims) -> grid
	# Pass out_dims instead of desc here and create desc in C code.

GpuDnnTransformerSampler(img, grid) -> transformed
	# out_dims can be inferred in C code (as we have img and grid).

GpuDnnTransformerGradI(img, grid, dy) -> dimg, dgrid
	# out_dims can be inferred in C code (as we have img and grid).

GpuDnnTransformerGradT(dgrid) -> dtheta
	# From dgrid we have N, H and W, but we miss C (number of channels).
	# Does this gradient really need to know C to compute dtheta ?
	# Could you try to use C == 1 to create the cuDNN desc, and check if it works ?
```

joaovictortr replied:
I've removed the descriptor from the implementation, now it is allocated inside the C code of each Op, following the comments on the interface from @notoraptor . In the case of ``GpuDnnTransformerGradT``, it is sufficient to set the number of channels in the descriptor to 1, because it is not a relevant information to that Op.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2739,240 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+            Affine transformation tensor containing one affine transformation
+            matrix per image. ``theta`` is usually generated by the localization
+            network.
+        """
```
Extend this docstring with details about `out_dims`.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2739,240 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+        return Apply(self, inputs, outputs)
+
+
+def dnn_spatialtf(img, theta, scale_width=1, scale_height=1, precision=theano.config.floatX):
```
Argument `precision` is now useless.

notoraptor commented on theano/gpuarray/dnn.py:
```diff
@@ -2738,6 +2739,240 @@ def dnn_batch_normalization_test(inputs, gamma, beta, mean, var,
+                theano.tensor.ceil(img.shape[3] * scale_width))
+    out_dims = tuple([as_scalar(v).astype('int64') for v in out_dims])
+
+    context_name = infer_context_name(img, theta)
```
You can remove lines from that line (2961) to line 2968, as GpuDnnTransformerGrid's and GpuDnnTransformerSampler's`make_node()`s already do these checkings (gpu_contiguous, as_gpu_array, ndim checking).

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_gi.c:
```diff
@@ -0,0 +1,194 @@
+            "GpuDnnTransformerGradI: failed to compute gradients of the inputs: %s",
+            cudnnGetErrorString( err ) );
+
+        return -1;
```
`return 1` instead here, as for the other error codes returned (just to make code consistent).

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_grid.c:
```diff
@@ -0,0 +1,129 @@
+
+#section init_code_struct
+
+cudnnStatus_t APPLY_SPECIFIC(err) = CUDNN_STATUS_SUCCESS;
```
You can put code into braces as you do in `dnn_sptf_gi.c`, so that you won't need to use `APPLY_SPECIFIC(err)`.

notoraptor commented on theano/gpuarray/c_code/dnn_sptf_gt.c:
```diff
@@ -0,0 +1,108 @@
+
+#section init_code_struct
+
+cudnnStatus_t APPLY_SPECIFIC(err) = CUDNN_STATUS_SUCCESS;
```
Here too (use braces).

joaovictortr commented:
Hello @notoraptor , thank you for reviewing the code again. I've addressed your comments in the latest commits.

notoraptor commented:
@abergeron Don't forget this PR!

notoraptor commented:
Don't forget a last pass on this PR @nouiz !
